internal/core/services/events_forwarder/events_forwarder_service.go

```diff
@@ -24,60 +24,169 @@ package events_forwarder
+	eventsForwarder ports.EventsForwarder,
+	schedulerStorage ports.SchedulerStorage,
+	instanceStorage ports.GameRoomInstanceStorage,
+	schedulerCache ports.SchedulerCache,
+) interfaces.EventsService {
 	return &EventsForwarderService{
 		eventsForwarder,
 		zap.L().With(zap.String("service", "rooms_api")),
+		schedulerStorage,
+		instanceStorage,
+		schedulerCache,
 	}
 }
 
 func (es *EventsForwarderService) ProduceEvent(ctx context.Context, event *events.Event) error {
-	room := &game_room.GameRoom{}
-	instance := &game_room.Instance{}
-
-	switch event.Name {
-	case events.RoomEvent:
-		return es.forwardRoomEvent(ctx, room, instance, map[string]interface{}{}, "")
-	case events.PlayerEvent:
-		return es.forwardPlayerEvent(ctx, room, map[string]interface{}{}, "")
+	var eventType string
+	if _, ok := event.Attributes["eventType"]; ok {
+		eventType = event.Attributes["eventType"].(string)
+	} else {
+		return errors.New("eventAttributes must contain key \"eventType\"")
 	}
-	return nil
 
+	var scheduler *entities.Scheduler
+	var err error
+	scheduler, _ = es.schedulerCache.GetScheduler(ctx, event.SchedulerID)
+	if scheduler == nil {
+		scheduler, err = es.schedulerStorage.GetScheduler(ctx, event.SchedulerID)
+		if err != nil {
+			es.logger.Error(fmt.Sprintf("Failed to get scheduler \"%v\" info", event.SchedulerID), zap.Error(err))
+			return err
+		}
+		if err := es.schedulerCache.SetScheduler(ctx, scheduler); err != nil {
+			es.logger.Error(fmt.Sprintf("Failed to set scheduler \"%v\" in cache", event.SchedulerID), zap.Error(err))
+		}
+	}
+
+	forwarderList := scheduler.Forwarders
+	if len(forwarderList) > 0 {
+		for _, _forwarder := range forwarderList {
+			switch event.Name {
+			case events.RoomEvent:
+				instance, err := es.instanceStorage.GetInstance(ctx, event.SchedulerID, event.RoomID)
+				if err != nil {
+					es.logger.Error(fmt.Sprintf("Failed to get instance for room \"%v\" from scheduler \"%v\" info", event.RoomID, event.SchedulerID), zap.Error(err))
+					return err
+				}
+				err = es.forwardRoomEvent(ctx, event, eventType, *instance, *scheduler, _forwarder)
+				if err != nil {
+					return err
+				}
+			case events.PlayerEvent:
+				err = es.forwardPlayerEvent(ctx, event, eventType, _forwarder)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		es.logger.Info(fmt.Sprintf("scheduler \"%v\" do not have forwarders configured", event.SchedulerID))
+	}
+
+	return nil
 }
 
-func (es *EventsForwarderService) forwardRoomEvent(ctx context.Context, room *game_room.GameRoom, instance *game_room.Instance, attributes map[string]interface{}, options interface{}) error {
-	err := es.eventsForwarder.ForwardRoomEventObsolete(ctx, room, instance, attributes, options)
+func (es *EventsForwarderService) forwardRoomEvent(
+	ctx context.Context,
+	event *events.Event,
+	eventType string,
+	instance game_room.Instance,
+	scheduler entities.Scheduler,
+	_forwarder *forwarder.Forwarder,
+) error {
+	var selectedPort int32
```

Comment:
@caiordjesus here I think you could switch conditions too like

if len(instance.Address.Ports) <= 0 {
		return errors.New("no room port found to forward roomEvent")	
	}
selectedPort := instance.Address.Ports[0].Port
….